File: Systems-for-Design-and-Implementation/Labs/Lab11/ProiectJava/TransportPersistence/src/main/java/transport/persistence/repository/ReservationDBRepository.java
package transport.persistence.repository;

import transport.model.Reservation;
import transport.model.Trip;
import transport.persistence.IReservationRepository;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Properties;

public class ReservationDBRepository implements IReservationRepository {

    private JdbcUtils dbUtils;
    //private static final Logger logger = LogManager.getLogger();

    public ReservationDBRepository(Properties props) {
        //logger.info("Initializing ReservationDBRepository with properties: {}", props);
        dbUtils = new JdbcUtils(props);
    }

    @Override
    public Optional<Reservation> findOne(Long id) {
        // method to find a reservation by id
        //logger.traceEntry();
        Connection con = dbUtils.getConnection();
        try (PreparedStatement preStmt = con.prepareStatement("select * from Reservations where id=?")) {
            preStmt.setLong(1, id);
            try (ResultSet result = preStmt.executeQuery()) {
                if (result.next()) {
                    // TRIP COLUMN
                    Long tripId = result.getLong("trip_id");
                    Trip trip = null;
                    try (PreparedStatement tripStmt = con.prepareStatement("select * from Trips where id=?")) {
                        tripStmt.setLong(1, tripId);
                        try (ResultSet tripResult = tripStmt.executeQuery()) {
                            if (tripResult.next()) {
                                String destination = tripResult.getString("destination");
                                Timestamp dateDeparture = tripResult.getTimestamp("dateDeparture");
                                Long noOfAvailableSeats = tripResult.getLong("noOfAvailableSeats");
                                trip = new Trip(destination, dateDeparture.toLocalDateTime(), noOfAvailableSeats);
                                trip.setId(tripId);
                            }
                        }
                    }
                    if (trip == null) {
                        return Optional.empty();
                    }

                    // RESERVEDSEATS COLUMN
                    List<Long> reservedSeats = new ArrayList<>();
                    try (PreparedStatement seatStmt = con.prepareStatement("select * from Seats where reservation_id=?")) {
                        seatStmt.setLong(1, id);
                        try (ResultSet seatResult = seatStmt.executeQuery()) {
                            while (seatResult.next()) {
                                reservedSeats.add(seatResult.getLong("seat_number"));
                            }
                        }
                    }

                    // CLIENTNAME COLUMN
                    String clientName = result.getString("client_name");

                    Reservation reservation = new Reservation(trip, reservedSeats, clientName);
                    reservation.setId(id);
                    return Optional.of(reservation);
                }
            }
        } catch (SQLException e) {
            //logger.error(e);
            System.out.println("Error DB " + e);
        }
        return Optional.empty();
    }

    @Override
    public Iterable<Reservation> findAll() {
        // method to find all reservations
        //logger.traceEntry();
        Connection con = dbUtils.getConnection();
        List<Reservation> reservations = new ArrayList<>();
        try (PreparedStatement preStmt = con.prepareStatement("select * from Reservations")) {
            try (ResultSet result = preStmt.executeQuery()) {
                while (result.next()) {
                    Long id = result.getLong("id");
                    Reservation reservation = findOne(id).orElse(null);
                    if (reservation != null) {
                        reservations.add(reservation);
                    }
                }
            }
        } catch (SQLException e) {
            //logger.error(e);
            System.out.println("Error DB " + e);
        }
        return reservations;
    }

    @Override
    public Optional<Reservation> update(Reservation reservation) {
        // method to update a reservation
        //logger.traceEntry("updating task {}", reservation);
        Connection con = dbUtils.getConnection();
        try {
            // Update the reservation in the Reservations table
            try (PreparedStatement preStmt = con.prepareStatement("update Reservations set trip_id=?, reserved_seats=?, client_name=? where id=?")) {
                preStmt.setLong(1, reservation.getTrip().getId());
                preStmt.setLong(2, reservation.getReservedSeats().size());
                preStmt.setString(3, reservation.getClientName());
                preStmt.setLong(4, reservation.getId());
                preStmt.executeUpdate();
            }

            // Delete all the old seats from the Seats table associated with the reservation
            try (PreparedStatement seatStmt = con.prepareStatement("delete from Seats where reservation_id=?")) {
                seatStmt.setLong(1, reservation.getId());
                seatStmt.executeUpdate();
            }

            // Insert the new seats in the Seats table
            for (Long seat : reservation.getReservedSeats()) {
                try (PreparedStatement newSeatStmt = con.prepareStatement("insert into Seats (reservation_id, seat_number) values (?, ?)")) {
                    newSeatStmt.setLong(1, reservation.getId());
                    newSeatStmt.setLong(2, seat);
                    newSeatStmt.executeUpdate();
                }
            }

            return Optional.of(reservation);
        } catch (SQLException e) {
            //logger.error(e);
            System.out.println("Error DB " + e);
        }
        return Optional.empty();
    }

    @Override
    public Optional<Reservation> save(Reservation reservation) {
        // method to save a reservation
        //logger.traceEntry("saving task {}", reservation);
        Connection con = dbUtils.getConnection();
        try (PreparedStatement preStmt = con.prepareStatement("insert into Reservations (trip_id, reserved_seats, client_name) values (?, ?, ?)", PreparedStatement.RETURN_GENERATED_KEYS)) {
            preStmt.setLong(1, reservation.getTrip().getId());
            preStmt.setLong(2, reservation.getReservedSeats().size());
            preStmt.setString(3, reservation.getClientName());
            preStmt.executeUpdate();

            try (Statement stmt = con.createStatement();
                 ResultSet rs = stmt.executeQuery("select last_insert_rowid()")) {
                if (rs.next()) {
                    reservation.setId(rs.getLong(1));

                    // Insert the seats in the Seats table
                    for (Long seat : reservation.getReservedSeats()) {
                        try (PreparedStatement seatStmt = con.prepareStatement("insert into Seats (reservation_id, seat_number) values (?, ?)")) {
                            seatStmt.setLong(1, reservation.getId());
                            seatStmt.setLong(2, seat);
                            seatStmt.executeUpdate();
                        }
                    }
                }
            }
        } catch (SQLException e) {
            //logger.error(e);
            System.out.println("Error DB " + e);
        }
        return Optional.of(reservation);
    }

    @Override
    public Optional<Reservation> delete(Long id) {
        // method to delete a reservation by id
        //logger.traceEntry("deleting task with {}", id);
        Connection con = dbUtils.getConnection();
        Optional<Reservation> reservation = findOne(id);
        if (reservation.isPresent()) {
            try {
                // Delete all the seats from the Seats table associated with the reservation
                try (PreparedStatement seatStmt = con.prepareStatement("delete from Seats where reservation_id=?")) {
                    seatStmt.setLong(1, id);
                    seatStmt.executeUpdate();
                }

                // Delete the reservation from the Reservations table
                try (PreparedStatement preStmt = con.prepareStatement("delete from Reservations where id=?")) {
                    preStmt.setLong(1, id);
                    preStmt.executeUpdate();
                }

                return reservation;
            } catch (SQLException e) {
                //logger.error(e);
                System.out.println("Error DB " + e);
            }
        }
        return Optional.empty();
    }
}

File: Systems-for-Design-and-Implementation/Labs/Lab4/ProiectCSharp/ProiectCSharp/repository/database/ReservationDBRepository.cs
using System.Data;
using log4net;
using ProiectCSharp.domain;

namespace ProiectCSharp.repository.database;

public class ReservationDBRepository
{
    private static readonly ILog logger = LogManager.GetLogger(typeof(ReservationDBRepository));
    private TripDBRepository tripRepository;
    
    public ReservationDBRepository()
    {
        logger.Info("Initializing ReservationDBRepository");
        tripRepository = new TripDBRepository();
    }

    public Reservation FindOne(long id)
    {
        logger.Info("Finding Reservation with id " + id);
        IDbConnection connection = DbUtils.getConnection();
        using (IDbCommand command = connection.CreateCommand())
        {
            command.CommandText = "SELECT * FROM Reservations WHERE id = @id";
            IDataParameter idParam = command.CreateParameter();
            idParam.ParameterName = "@id";
            idParam.Value = id;
            command.Parameters.Add(idParam);
            using (IDataReader result = command.ExecuteReader())
            {
                if (result.Read())
                {
                    // TRIP COLUMN
                    long tripId = result.GetInt64(1);
                    Trip trip = tripRepository.FindOne(tripId);
                    if (trip == null)
                    {
                        return null;
                    }

                    // RESERVEDSEATS COLUMN
                    List<long> reservedSeats = new List<long>();
                    using (IDbCommand seatCommand = connection.CreateCommand())
                    {
                        seatCommand.CommandText = "SELECT * FROM Seats WHERE reservation_id = @id";
                        IDataParameter seatIdParam = seatCommand.CreateParameter();
                        seatIdParam.ParameterName = "@id";
                        seatIdParam.Value = id;
                        seatCommand.Parameters.Add(seatIdParam);
                        using (IDataReader seatResult = seatCommand.ExecuteReader())
                        {
                            while (seatResult.Read())
                            {
                                reservedSeats.Add(seatResult.GetInt64(0));
                            }
                        }
                    }
                    
                    // CLIENTNAME COLUMN
                    string clientName = result.GetString(3);
                    Reservation reservation = new Reservation(trip, reservedSeats, clientName);
                    reservation.Id = id;
                    return reservation;
                }
            }
        }
        return null;
    }
    
    public IEnumerable<Reservation> FindAll()
    {
        logger.Info("Finding all Reservations");
        IDbConnection connection = DbUtils.getConnection();
        List<Reservation> reservations = new List<Reservation>();
        using (IDbCommand command = connection.CreateCommand())
        {
            command.CommandText = "SELECT * FROM Reservations";
            using (IDataReader result = command.ExecuteReader())
            {
                while (result.Read())
                {
                    long id = result.GetInt64(0);
                    Reservation reservation = FindOne(id);
                    if (reservation != null)
                    {
                        reservations.Add(reservation);
                    }
                }
            }
        }
        return reservations;
    }
    
    public Reservation Save(Reservation reservation)
    {
        logger.Info("Saving Reservation " + reservation);
        IDbConnection connection = DbUtils.getConnection();
        using (IDbCommand command = connection.CreateCommand())
        {
            command.CommandText = "INSERT INTO Reservations (trip_id, reserved_seats, client_name) VALUES (@trip_id, @reserved_seats, @client_name)";
            IDataParameter tripIdParam = command.CreateParameter();
            tripIdParam.ParameterName = "@trip_id";
            tripIdParam.Value = reservation.Trip.Id;
            command.Parameters.Add(tripIdParam);
            IDataParameter seatsParam = command.CreateParameter();
            seatsParam.ParameterName = "@reserved_seats";
            seatsParam.Value = reservation.ReservedSeats.Count;
            command.Parameters.Add(seatsParam);
            IDataParameter clientNameParam = command.CreateParameter();
            clientNameParam.ParameterName = "@client_name";
            clientNameParam.Value = reservation.ClientName;
            command.Parameters.Add(clientNameParam);
            command.ExecuteNonQuery();

            // Get the last inserted id
            command.CommandText = "SELECT last_insert_rowid()";
            long lastId = (long)command.ExecuteScalar();
            reservation.Id = lastId;

            // Insert the seats in the Seats table
            foreach (long seat in reservation.ReservedSeats)
            {
                using (IDbCommand seatCommand = connection.CreateCommand())
                {
                    seatCommand.CommandText = "INSERT INTO Seats (reservation_id, seat_number) VALUES (@reservation_id, @seat_number)";
                    IDataParameter seatIdParam = seatCommand.CreateParameter();
                    seatIdParam.ParameterName = "@reservation_id";
                    seatIdParam.Value = reservation.Id;
                    seatCommand.Parameters.Add(seatIdParam);
                    IDataParameter seatNumberParam = seatCommand.CreateParameter();
                    seatNumberParam.ParameterName = "@seat_number";
                    seatNumberParam.Value = seat;
                    seatCommand.Parameters.Add(seatNumberParam);
                    seatCommand.ExecuteNonQuery();
                }
            }
        }
        return reservation;
    }
    
    public Reservation Delete(long id)
    {
        logger.Info("Deleting Reservation with id " + id);
        Reservation reservation = FindOne(id);
        if (reservation != null)
        {
            IDbConnection connection = DbUtils.getConnection();
            // Delete all the seats from the Seats table associated with the reservation
            using (IDbCommand seatCommand = connection.CreateCommand())
            {
                seatCommand.CommandText = "DELETE FROM Seats WHERE reservation_id = @id";
                IDataParameter seatIdParam = seatCommand.CreateParameter();
                seatIdParam.ParameterName = "@id";
                seatIdParam.Value = id;
                seatCommand.Parameters.Add(seatIdParam);
                seatCommand.ExecuteNonQuery();
            }

            // Delete the reservation from the Reservations table
            using (IDbCommand command = connection.CreateCommand())
            {
                command.CommandText = "DELETE FROM Reservations WHERE id = @id";
                IDataParameter idParam = command.CreateParameter();
                idParam.ParameterName = "@id";
                idParam.Value = id;
                command.Parameters.Add(idParam);
                command.ExecuteNonQuery();
            }
        }
        return reservation;
    }
    
    public Reservation Update(Reservation reservation)
    {
        logger.Info("Updating Reservation " + reservation);
        IDbConnection connection = DbUtils.getConnection();
        using (IDbCommand command = connection.CreateCommand())
        {
            command.CommandText = "UPDATE Reservations SET trip_id = @trip_id, reserved_seats = @reserved_seats, client_name = @client_name WHERE id = @id";
            IDataParameter tripIdParam = command.CreateParameter();
            tripIdParam.ParameterName = "@trip_id";
            tripIdParam.Value = reservation.Trip.Id;
            command.Parameters.Add(tripIdParam);
            IDataParameter seatsParam = command.CreateParameter();
            seatsParam.ParameterName = "@reserved_seats";
            seatsParam.Value = reservation.ReservedSeats.Count;
            command.Parameters.Add(seatsParam);
            IDataParameter clientNameParam = command.CreateParameter();
            clientNameParam.ParameterName = "@client_name";
            clientNameParam.Value = reservation.ClientName;
            command.Parameters.Add(clientNameParam);
            IDataParameter idParam = command.CreateParameter();
            idParam.ParameterName = "@id";
            idParam.Value = reservation.Id;
            command.Parameters.Add(idParam);
            command.ExecuteNonQuery();
        }

        // Delete all the old seats from the Seats table associated with the reservation
        using (IDbCommand seatCommand = connection.CreateCommand())
        {
            seatCommand.CommandText = "DELETE FROM Seats WHERE reservation_id = @id";
            IDataParameter seatIdParam = seatCommand.CreateParameter();
            seatIdParam.ParameterName = "@id";
            seatIdParam.Value = reservation.Id;
            seatCommand.Parameters.Add(seatIdParam);
            seatCommand.ExecuteNonQuery();
        }

        // Insert the new seats in the Seats table
        foreach (long seat in reservation.ReservedSeats)
        {
            using (IDbCommand newSeatCommand = connection.CreateCommand())
            {
                newSeatCommand.CommandText = "INSERT INTO Seats (reservation_id, seat_number) VALUES (@reservation_id, @seat_number)";
                IDataParameter newSeatIdParam = newSeatCommand.CreateParameter();
                newSeatIdParam.ParameterName = "@reservation_id";
                newSeatIdParam.Value = reservation.Id;
                newSeatCommand.Parameters.Add(newSeatIdParam);
                IDataParameter seatNumberParam = newSeatCommand.CreateParameter();
                seatNumberParam.ParameterName = "@seat_number";
                seatNumberParam.Value = seat;
                newSeatCommand.Parameters.Add(seatNumberParam);
                newSeatCommand.ExecuteNonQuery();
            }
        }
        return reservation;
    }
}